src/codegen/operator/csv_scan_translator.cpp

```diff
@@ -0,0 +1,307 @@
+#include "codegen/lang/if.h"
+#include "codegen/operator/projection_translator.h"
+#include "codegen/pipeline.h"
+#include "codegen/proxy/csv_scanner_proxy.h"
+#include "codegen/proxy/runtime_functions_proxy.h"
+#include "codegen/type/sql_type.h"
+#include "codegen/vector.h"
+#include "planner/csv_scan_plan.h"
+
+namespace peloton {
+namespace codegen {
+
+CSVScanTranslator::CSVScanTranslator(const planner::CSVScanPlan &scan,
+                                     CompilationContext &context,
+                                     Pipeline &pipeline)
+    : OperatorTranslator(scan, context, pipeline) {
+  // Register the CSV scanner instance
+  auto &query_state = context.GetQueryState();
+  scanner_id_ = query_state.RegisterState(
+      "csvScanner", CSVScannerProxy::GetType(GetCodeGen()));
+
+  // Load information about the attributes output by the scan plan
+  scan.GetAttributes(output_attributes_);
+}
+
+void CSVScanTranslator::InitializeQueryState() {
+  auto &codegen = GetCodeGen();
+
+  auto &scan = GetPlanAs<planner::CSVScanPlan>();
+
+  // Arguments
+  llvm::Value *scanner_ptr = LoadStatePtr(scanner_id_);
+  llvm::Value *exec_ctx_ptr = GetExecutorContextPtr();
+  llvm::Value *file_path = codegen.ConstString(scan.GetFileName(), "filePath");
+
+  auto num_cols = static_cast<uint32_t>(output_attributes_.size());
+
+  // We need to generate an array of type::Type. To do so, we construct a vector
+  // of the types of the output columns, and we create an LLVM constant that is
+  // a copy of the underlying bytes.
+
+  std::vector<codegen::type::Type> col_types_vec;
+  col_types_vec.reserve(num_cols);
+  for (const auto *ai : output_attributes_) {
+    col_types_vec.push_back(ai->type);
+  }
+  llvm::Value *raw_col_type_bytes = codegen.ConstGenericBytes(
+      col_types_vec.data(), static_cast<uint32_t>(col_types_vec.capacity()),
+      "colTypes");
+  llvm::Value *output_col_types = codegen->CreatePointerCast(
+      raw_col_type_bytes, TypeProxy::GetType(codegen)->getPointerTo());
+
+  // Now create a pointer to the consumer function
+  using ConsumerFuncType = void (*)(void *);
+  llvm::Value *consumer_func = codegen->CreatePointerCast(
+      consumer_func_, proxy::TypeBuilder<ConsumerFuncType>::GetType(codegen));
+
+  // Cast the runtime type to an opaque void*. This is because we're calling
+  // into pre-compiled C++ that doesn't know that the dynamically generated
+  // RuntimeState* looks like.
+  llvm::Value *query_state_ptr =
+      codegen->CreatePointerCast(codegen.GetState(), codegen.VoidPtrType());
+
+  // Call CSVScanner::Init()
+  codegen.Call(CSVScannerProxy::Init,
+               {scanner_ptr, exec_ctx_ptr, file_path, output_col_types,
+                codegen.Const32(num_cols), consumer_func, query_state_ptr,
+                codegen.Const8(scan.GetDelimiterChar()),
+                codegen.Const8(scan.GetQuoteChar()),
+                codegen.Const8(scan.GetEscapeChar())});
+}
+
+namespace {
```

Comment:
?

Reply:
This is just an anonymous namespace. I use it so I can create classes in `.cpp` files without worrying about name collisions.